content/blog/2016/2016-08-29-jenkins-world-speaker-blog-goodgame.adoc

```diff
@@ -0,0 +1,42 @@
+
+Of course the needs depend strongly on the way Jenkins is being used, so not every rule implemented might be useful for every team, but this applies as well all other code-analysis. The main inspiration for the rules were developer feedback and some articles found in the web. The different possibilities to use and configure Jenkins provides a lot of potential for many more rules. With this new approach of quality-analysis, we can enforce best practices like:
+
+* Polling must die (Rather trigger a build due to pushes instead of poll the repository every x minutes)
+* Use Log Rotator (Not using log-rotator can result in disk space problems on the master)
+* Use slaves/labels (Jobs should be defined where to run)
+* Don’t build on the master (In larger systems, don't build on the master)
+* Enforce plugin usage (For example: Timestamp, Mask-Passwords)
+* Naming sanity (Limit project names to a sane (e.g. alphanumeric) character set)
+* Analyze Groovy Scripts (For example: Prevent System.exit(0) in System Groovy Scripts)
+
+image:/images/post-images/jw-speaker-blog-goodgame/jenkins1.png[role=left]
+
+Besides taking control over all configuration of any Jenkins-instance we want, there is also room for additional metrics, like measuring the amount and different types of jobs (Freestyle/Maven etc...) to get an overview about the general load of the Jenkins-instance. A more sophisticated idea is to measure complexity of jobs and even pipelines. As code, jobs-configuration gets harder to understand the more steps are involved. On the one hand scripts, conditions and many parameters can negatively influence the readability, especially if you have external dependencies (like scripts) in different locations. On the other hand, pipelines can also grow very complex when many jobs are involved and chained for execution. It will be very interesting for us too see where and why too complex pipelines are being created.
+
+On visualization we rely on the data and its interpretation of SonarQube, which offers a big bandwidth of widgets. Everybody can use and customize the dashboards. Our centralized team for example has a separate dashboard where we can get a quick overview over all instances.
+
+image:/images/post-images/jw-speaker-blog-goodgame/jenkins2.png[role=left]
+
+The problem of "growing" Jenkins with maintenance problems is not new. Especially when you have many developers involved, including with the access to create jobs and pipelines themselves, an analysis like this SonarQube-plugin provides can be useful for anyone who wants to keep their Jenkins in shape. Customization and standards are playing a big role in this scenario. This talk surely is not an advertisement for my developed plugin, it is more about the crazy idea using static code-analysis for Jenkins-job-configuration. I haven’t seen anything like it so far and I feel that there might me some potential behind this idea.
```

Comment:
"SonarQube plugin"
s/This talk/This blog post/
"crazy idea of using"
"code analysis"
"Jenkins job configuration"
s/might me/might be/